### DISCUSSION TITLE
A big thank you for the recent Arrow Keys integration!

### DISCUSSION THREAD

**Discussion by SHHSSH** [piroor/treestyletab#3574] [General]
Thank you @piroor for the recent update of simulating arrow keys for navigation in TST, this has been immensely beneficial for utilisation with TST in a Separate Window helper addon. 

Although for the functionality of it to be 100% and actually be able to Left & Right arrow keys in TST separate window, I must still have TST sidebar tree active, I cannot toggle its visibility to not visible.

This is able to be circumvented via userChrome.css nonetheless with a minimum width for the sidebar and I just set it to 2 and manually drag TST back to visible if I need to manually use it with the mouse :)

I mirror the TST in Separate Window window with OnTopReplica & set it to AlwaysOnTop and nircmd its transparency to 0.
This allows me to have a fullscreen Firefox for better immersion and a customisable overlay for less mess which I can easily also then toggle AlwaysOnTop which a simple mouse macro of the Firefox window for toggling visibility of the overlays.

**Thank you very, very much!! Your work is truly appreciated!** 
It's actually crazy I was just about to request this feature within the past 48 hours for this exact reason.
I am not very good with actual programming but I find myself fiddling with a lot of gadgets and your Add-on has been such a great browsing tool over the years.
- A few random screenshots
With the usage of certain macro scripts, I am able to basically browse entirely with my mouse now.
If anyone cares, the functionality of the keys on my mouse are able to be multiplied by the Ctrl+Alt+Shift+Win keys on the mouse too, so essentially you can have quadruple the macros to use to your preference.

![capture_088_12062024_052434](https://github.com/piroor/treestyletab/assets/21285581/105b9d65-ab20-4559-ae3c-2be82f61099d)
![capture_099_12062024_055747](https://github.com/piroor/treestyletab/assets/21285581/4ba34a65-dbda-4643-9ffd-4f92a45c2a25)
![capture_091_12062024_052756](https://github.com/piroor/treestyletab/assets/21285581/a110a069-ad7c-4f4c-8c97-f3c719b2c76a)
![capture_093_12062024_052843](https://github.com/piroor/treestyletab/assets/21285581/2c952d07-47d0-454c-9da9-27b44f7a5a5f)
![capture_094_12062024_052857](https://github.com/piroor/treestyletab/assets/21285581/b21f89fa-807a-421f-8684-e04aafcdf48f)
![capture_095_12062024_052907](https://github.com/piroor/treestyletab/assets/21285581/160bb752-82cf-4afc-b058-a6877c973c55)

*EDIT* Nowadays I just macro a transparency hotkey to an AHK script - 

#NoEnv  ; Recommended for performance and compatibility with future AutoHotkey releases.
#Warn  ; Enable warnings to assist with detecting common errors.
SendMode Input  ; Recommended for new scripts due to its superior speed and reliability.
SetWorkingDir %A_ScriptDir%  ; Ensures a consistent starting directory.

; Run the script as administrator
if not A_IsAdmin
{
   Run *RunAs "%A_ScriptFullPath%"  ; Requires v1.0.92.01+
   ExitApp
}

^F7::  ; Ctrl+F7 to decrease opacity (make more transparent)
    Send !+{F5}  ; Send Alt+Shift+F5
    WinGetActiveTitle, activeWindow
    if (!InStr(activeWindow, "Obsidian"))
    {
        WinGet, active_id, ID, A
        WinGet, trans, Transparent, ahk_id %active_id%
        if (trans = "")
            trans := 255
        trans -= 100
        if (trans < 0)
            trans := 0
        WinSet, Transparent, %trans%, ahk_id %active_id%
    }
return

^F8::  ; Ctrl+F8 to increase opacity (make less transparent)
    Send !+{F12}  ; Send Alt+Shift+F12
    WinGetActiveTitle, activeWindow
    if (!InStr(activeWindow, "Obsidian"))
    {
        WinGet, active_id, ID, A
        WinGet, trans, Transparent, ahk_id %active_id%
        if (trans = "")
            trans := 0
        trans += 100
        if (trans > 255)
            trans := 255
        WinSet, Transparent, %trans%, ahk_id %active_id%
    }
return

**Comment by SHHSSH** [comment]
Is there any possible solution for the malfunction of Down and Right Arrow keys when TST Sidebar is not active? Only seems that Left and Up work and if I press Right or Down Arrow keys it malfunctions and goes to the first tab.

This would be amazing if this was rectifiable as I often open an additional webpage in my sidebar and cannot anymore.

**Reply by piroor** [reply]
Unstable behavior with closed sidebar is a bug of TST, sorry. I've pushed a fix so those shortcuts will fallback to "focus to next tab" with closed sidebar at the next release.

**Reply by SHHSSH** [reply]
> Unstable behavior with closed sidebar is a bug of TST, sorry. I've pushed a fix so those shortcuts will fallback to "focus to next tab" with closed sidebar at the next release.

Right should fallback to extending of a tree too(as it does when TST sidebar is active), I think focus to next tab is maybe only relevant for Down arrow key.
As with Left arrow key I can collapse a tree.

**Reply by piroor** [reply]
Currently the "fallback" happens when the sidebar is closed and you cannot see the state of collapsed/expanded. Should Left/Right shortcuts collapse/expand tree even if the sidebar is closed? (I'm afraid of users' confusion: when Left key shortcut is triggered on a child tab, the user will see odd behaviors like "nothing happened for the first input (it actually collapsed the tree internally but you can't see that), and the left tab is focused for the second input".)

**Reply by SHHSSH** [reply]
I understand what you're saying. 
Maybe an alternative/additional hotkey option for those who may want to use the Arrow keys' navigation in a different manner.
Potentially via independent hotkeys for "Simulate Left Key on Parent Tab" & "Simulate Right Key on Parent Tab" would be an appropriate solution? Preferably these new functions STILL work the same as Left & Right Arrow key navigation and the originals stay the same, however, if the user is using "Simulate Left/Right Key on Parent Tab" those functions also have the capability to Collapse/Expand Parent Tab Tree.
_So yes, even though both hotkey options have the capability of collapsing/expanding(when TST is **open**) with original Left/Right, with an **independent** hotkey for Parent Tab Left/Right simulation it allows for the collapsing/expanding when TST is **closed** so that the user is aware of what's occurring and the general user will not have to worry about a "nothing happened" command._

Original Down/Right Arrow keys' fallback when TST is closed to "focus to next tab" is still a good solution for Down/Right Arrow keys' malfunction with TST sidebar closed though(especially for general users).
As even when the tree is not locked as collapsed, it still skips over the tree with Up/Down Arrow key navigation.
But I don't believe LAC helper addon is necessarily even relevant to this discussion, so I'm sorry I keep referencing it sometimes.

**Reply by piroor** [reply]
I'm very sorry I missed that you use the separate window addon. I used wrong utility method which detect the in-window sidebar is really opened instead of any sidebar panel for the window is opened, to switch the behavior of keyboard shortcuts. The next release including the change 4d3e1994904501a650de07df26cb008aaa6a02e9 will work better for your case.

**Reply by SHHSSH** [reply]
> I'm very sorry I missed that you use the separate window addon. I used wrong utility method which detect the in-window sidebar is really opened instead of any sidebar panel for the window is opened, to switch the behavior of keyboard shortcuts. The next release including the change [4d3e199](https://github.com/piroor/treestyletab/commit/4d3e1994904501a650de07df26cb008aaa6a02e9) will work better for your case.

I got it working dude, all is well.

**Comment by SHHSSH** [comment]
You are the King and probably already know how to do all that kind of stuff yourself but if you maybe need help with the parent tab identification or anything relevant to developing the workaround with my suggestion perhaps you could check out this Addon's features - https://addons.mozilla.org/en-US/firefox/addon/tst-number-navigation/
![capture_006_13062024_015351](https://github.com/piroor/treestyletab/assets/21285581/94baad13-7e97-4b03-8f03-07a47c243829)

**Comment by SHHSSH** [comment]
I didn't notice but you already have similar hotkeys.
I am going to fiddle with this for over the next few days to see if any helps :)

***EDIT*** - For anyone in the future maybe looking to solve this, currently this setup is working well with TST Sidebar _closed_
Obviously you just set your preferred hotkeys.
![capture_007_13062024_185710](https://github.com/piroor/treestyletab/assets/21285581/2bb8554f-f903-4163-ac53-759f8cebd429)

Thanks @piroor I think this actually already solves it, disregard the other considerations, sorry for the waste of time.

**Comment by preetmyob** [comment]
Adding this by default has made using those keystrokes in text boxes for selecting and editing text by keyboard shortcut that works everywhere else non-functional. When I went into check I saw that there were 49 keyboard shortcuts enabled. I'm not sure if all or just some are new, so I've had to switch off the extension. I think doing this by default is probably not a great idea.

This is a big thing for people doing data analysis in Jupyter notebooks.

**Reply by SHHSSH** [reply]
Yes, I would not suggest using _actual_ arrow keys as the hotkeys for the simulation of them.
Therefore, perhaps you can just set a random unused hotkey for them to prevent that issue you're having from occurring.

**Reply by SHHSSH** [reply]
@piroor Just a backtracking reference for you, I agree that for sake of better uniformity, you should probably still consider having this off by default, though I personally use it as a primary form of navigation now.

**Reply by piroor** [reply]
Hmm, should I undefine default shortcuts and suggest users to activate them dynamically with https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/commands/update at the initial startup?

**Reply by SHHSSH** [reply]
> Hmm, should I undefine default shortcuts and suggest users to activate them dynamically with https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/commands/update at the initial startup?

I haven't installed a fresh TST without imported settings etc. nor on a newer profile of FF in a while, so I actually was(for whatever reason) under the impression hotkeys weren't at all defaulted "On" in any sense? At least not in the hotkeys of FF's Addon's Manager.

You're speaking regarding internally of TST? Either way, yes, I think it should be defaulted to off, primarily for more complex or unique features like this, other hotkeys(basic ones etc.) should perhaps be kept on by default.

**Comment by susieswallows** [comment]
Hi Mate, 

I realise this is probably a long ass time since you had this issue - but do you think your solution could help me out with my issue? 
It seems from what I've read, TST cannot be disabled... I'm not sure if this is DEFINITELY the case but looks to be so 😔

My forum is here in QnA: https://github.com/piroor/treestyletab/discussions/3627

Do let me know if you think your issue will fix it (if you still have access to have notifications/emails from GitHub lol), if you do get back to me and your solution does work, I'll def follow your tutorial. 

Thankies!

Susie

**Reply by SHHSSH** [reply]
> Hi Mate,
> 
> I realise this is probably a long ass time since you had this issue - but do you think your solution could help me out with my issue? It seems from what I've read, TST cannot be disabled... I'm not sure if this is DEFINITELY the case but looks to be so 😔
> 
> My forum is here in QnA: #3627
> 
> Do let me know if you think your issue will fix it (if you still have access to have notifications/emails from GitHub lol), if you do get back to me and your solution does work, I'll def follow your tutorial.
> 
> Thankies!
> 
> Susie

Hi Susie,

To clarify my macros(hotkeys) are primarily for navigation throughout the vertical tree(TST) and the repositioning is done manually with these macros. I do personally often sort into "folders" which are just blank tabs that index their sub-tabs/trees, however, this approach of mine doesn't particularly "sort" automatically on any basis, this is specifically to enhance my speed of moving tabs around TST without the need for the keyboard as much.

Regarding the following question of yours - " Is there a way I can collapse all the different family tabs in their respective trees from the top where the tabs usually are?";

I don't really understand what you mean by this, are you looking for an icon of sorts or something that would be capable of such? I don't believe there's such a thing, but you can use many commands of TST through its toolbar icon.

If you'd like any further clarification or a helping hand with anything, just ask.
